Refactor init for Excel readers to _BaseExcelReader

tdamsma opened the pull request:
Pre-cursor for #25092. To support multiple excel readers the init function is moved to _BaseExcelReader. The function is made a bit more general to be able to support multiple excel readers.

- [ ] tests passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

tdamsma committed:
Attempt to generalize _XlrdReader __init__ and move it to _BaseExcelReader

tdamsma committed:
Merge remote-tracking branch 'upstream/master' into excel-read-shared-init-to-baseclass

# Conflicts:
#	pandas/io/excel/_base.py
#	pandas/io/excel/_xlrd.py

tdamsma committed:
Merge branch 'master' into excel-read-shared-init-to-baseclass

jreback commented:
looks good. ping when ready. cc @WillAyd

WillAyd commented on pandas/io/excel/_xlrd.py:
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
-                try:
-                    # GH 19779
-                    filepath_or_buffer.seek(0)
-                except UnsupportedOperation:
```
I see this wasn't carried over - is it no longer valid?

tdamsma replied:
This is now dealt with by wrapping url responses in `BytesIO(urlopen(filepath_or_buffer).read())`. So now all items with `"read"` can also support `seek(0)`

WillAyd commented on pandas/io/excel/_xlrd.py:
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
-                    # GH 20434
-                    pass
+        self._engine = xlrd
+        super(_XlrdReader, self).__init__(filepath_or_buffer)
```
We just switched to 3.0 syntax so just call `super()` here

tdamsma replied:
agreed, implemented

WillAyd commented on pandas/io/excel/_xlrd.py:
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
+
+    @property
+    def _workbook_class(self):
+        return self._engine.Book
```
I think it would be better if this was namespaced directly from xlrd instead of going through `self._engine`

tdamsma replied:
agreed, implemented

tdamsma committed:
implement suggestions  @WillAyd

tdamsma committed:
Merge remote-tracking branch 'upstream/master' into excel-read-shared-init-to-baseclass

tdamsma committed:
remove _engine keyword altogether

tdamsma commented:
@jreback, @WillAyd, I don't understand the test failing tests, but other than that I think it's ready

WillAyd commented:
@tdamsma restarted all of them. Not sure what is wrong with the checks_and_doc job off the top of my head but looks like this PR introduced some regressions with handling of S3 excel files as shown in the Travis failures. 

If you could take a look would be great

tdamsma committed:
fix regression for reading s3 files

jreback commented:
lgtm. @WillAyd

tdamsma commented:
@WillAyd, I addressed the regression. Also I accidentally committed (and reverted) a setting to limit the test scope, that explains the codecov report.

WillAyd commented:
Nice job @tdamsma - looking forward to the follow ups!

tdamsma commented:
Great, thanks for the reviews
